% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/02_constructors.R
\name{ProteinCodingGene}
\alias{ProteinCodingGene}
\title{Constructor function for ProteinCodingGene class}
\usage{
ProteinCodingGene(ID, HUGO_symbol, gene_name, description, gene_structure, protein_ID, protein_sequence)
}
\arguments{
\item{ID}{The gene ID (Ensembl ID or NCBI gene ID).}

\item{HUGO_symbol}{The HUGO gene symbol.}

\item{gene_name}{The full gene name.}

\item{description}{A brief description of the gene.}

\item{gene_structure}{The gene structure represented as a GRanges object.}

\item{protein_ID}{The protein ID.}

\item{protein_sequence}{The protein sequence.}
}
\value{
An instance of the ProteinCodingGene class.
}
\description{
This function creates a new instance of the ProteinCodingGene class.
}
\examples{
gene_structure <- GenomicRanges::GRanges(seqnames = "chr1", ranges = IRanges::IRanges(1000, 2000), strand = "+")
ProteinCodingGene(ID = "ENSG000001", HUGO_symbol = "BRCA1",
                  gene_name = "Breast cancer 1",
                  description = "Breast cancer type 1 susceptibility protein",
                  gene_structure = gene_structure,
                  protein_ID = "ENSP000003", protein_sequence = "MENSDRNSIKVAV...")
}
\references{
\url{https://www.ensembl.org/index.html?redirect=no}\cr
}
\author{
Niccolò Bianchi\cr UniMi + PoliMI\cr Maintainer: Niccolò
Bianchi\cr E-Mail: \href{mailto:niccolo.bianchi2@studenti.unimi.it}{niccolo.bianchi2@studenti.unimi.it}
}
